internal/db/update/schema.go

```diff
@@ -153,6 +153,15 @@ func ensureUpdatesAreApplied(ctx context.Context, tx *sql.Tx, versions []int, sc
 			return nil
 		}
 
+		// Ensure we turn off foreign keys so any external table we don't know about
```

Comment:
If one of the external tables columns has a foreign key on any of the internal tables, how would this affect the relation? 
I am wondering if this could lead to an inconsistent state in the external tables as a user of the external table would expect the foreign key constraint to be working.

Wouldn't it be better to allow this behavior but make it clear that you should never reference those tables as they are internal?

Reply:
It shouldn't matter since the foreign keys are turned back on after the internal updates are applied. Since all internal updates are applied before external ones, I don't see any way that this would interfere with external tables.

That being said, it appears my approach was a bit naive because you can't set `PRAGMA foreign_keys=OFF` from within a transaction, which means this PR is going to have to be a bit more involved.

Reply:
Take the following as an example:

```sql
CREATE TABLE internal (
  column1 INTEGER PRIMARY KEY
)

CREATE TABLE external (
  column1 INTEGER
  FOREIGN KEY (column1) REFERENCES "internal" (column1)
)
```

Now let's assume there is a row in `external` which points to another row in `internal`. If we do `PRAGMA foreign_keys=OFF` (or something equal that works) and get rid of the row in `internal` and later on turn foreign keys back on, wouldn't the foreign key constraint `external(column1) -> internal(column1)` be broken?

Reply:
Yes it would, however there are no pre-existing rows in any table, they are all determined based on user input so there would never be a valid schema update that purposely deletes a row from a table, that would be extremely dangerous. 


If we do for some reason have to delete a row, it makes way more sense to me to actually break the relation first because we shouldn't be arbitrarily deleting data for an external schema. At least in that case the external schema has some semblance of pathway for recovery, instead of having rows in its tables deleted, they would simply have dereferenced rows. 

In the present case, we are migrating rows to a different table which we then rename back to the old table name. This is a common occurrence for schema updates, but if we don't turn foreign keys off first each time, then when we drop the original table, any external tables with a foreign key to that one will be wiped. And since we we can't turn primary keys off in a transaction that means we can't run all updates in one big transaction like we do now either.

Reply:
Since external schema updates are all run after the internal ones, these updates can be used to reconcile the external schema with the new internal one if the rows continue to exist in the external table. I can't think of any situation where we would be okay with deleting data external to microcluster from an internal schema update.

Reply:
It's clear to me that the internal schema update should never do any modifications to any of the external schemas/data. 

However it feels like a huge burden on our end as we cannot modify the schema (e.g. rename rows) without fearing that this might break an external setup with active foreign keys on those rows. Maybe the example from before wasn't ideal, but what if we decide to change `column1` to `column2` in the updated `internal` table?

As an appendix to your list from below, we could provide views to export a defined set of data. This won't allow foreign key constraints (as rows in views cannot be referenced by foreign key) but it could be used to reference the internal data and we are always able to modify the underlying structure using internal schema updates.

Reply:
Couple of points/questions here:

1. Once we have the LTS branch of microcluster then we cannot make *any* schema changes inside that branch after the first release (this is our commitment to allow downgrades).
2. Is there anything stopping microcluster have its own internal core DB, and the microcluster consumers (microcloud, microovn, microceph etc) having their own app specific database, and accessing microcluster core data via an API. This would prevent cross-db references?

Reply:
> On 2., I don't see how that solves the problem posed in step 1. It's just punting it to another area. If the upstream project relies on the API to balance its own database then we would be just as locked out of making API changes as we are currently for schema changes, because there's no reverting the effect that the API has on the upstream database once the nodes start up for the first time and rebalance.
> 
> On 3. Yes, the schema was designed to be as minimal as possible such that the only situations where we would need a schema update are when we are adding some feature (in this case: API extensions). We ended up with a second schema update from #94 because we found a bug with the schema update process that locked only the internal side out from adding such features. We offer a table for join tokens that just store each token's name and secret key, and a table for cluster members. External projects are expected to reference these tables and our schema updates should only be to extend the data we offer here.

I think this philosophy fits with point 1 because we won't be continually adding features to an LTS release.

Reply:
> On 2., I don't see how that solves the problem posed in step 1. It's just punting it to another area. If the upstream project relies on the API to balance its own database then we would be just as locked out of making API changes as we are currently for schema changes, because there's no reverting the effect that the API has on the upstream database once the nodes start up for the first time and rebalance.

The way I see that it solves/clarifies the problem as it defines clearly what is microcluster's API commitments to its consumers, rather than having an a potential unknown dependency on its own internal core tables.

If you are saying that the microcluster core internal tables *are* also an API, then it becomes quite simple really, we just cannot change the schema in a breaking way.

Reply:
TBH I am very confused by this now, will wait until next week when we can discuss in person.

Reply:
> If you are saying that the microcluster core internal tables _are_ also an API, then it becomes quite simple really, we just cannot change the schema in a breaking way.

Yes, this is a good way to sum up my thoughts on this. The internal database is simply another API that microcluster provides, and since it's a database we don't have to implement all the mechanics that interconnect the upstream database to the data we supply, we can just rely on sqlite handling it for us.